### ISSUE TITLE
Server closed the connection unexpectedly

### ISSUE THREAD

**Issue by ghost** [celery/django-celery#46]
I'm having a weird issue with PostgreSQL (8.4.7) and celery (2.2.5) and Django (1.2.3).

Suddenly all my celery tasks, that are touching the database, are _randomly_ raising:

```
DatabaseError: server closed the connection unexpectedly
    This probably means the server terminated abnormally
    before or while processing the request.
```

Sometimes a task fails 3 times, and then works just fine, other times it fails randomly.

if I run:

```
$python manage.py shell
```

and create an object and save it to the database, all works fine.
Here's the traceback:

```
[2011-04-14 16:59:43,071: INFO/MainProcess] Got task from broker:   WebTools.controlpanel.tasks.discoverDeviceTask[15dcea93-d3ff-416d-aedd-bc6fe80a6f6b]
[2011-04-14 16:59:43,122: WARNING/PoolWorker-25] server closed the connection unexpectedly
    This probably means the server terminated abnormally
    before or while processing the request.
[2011-04-14 16:59:43,126: ERROR/MainProcess] Task WebTools.controlpanel.tasks.discoverDeviceTask[15dcea93-d3ff-416d-aedd-bc6fe80a6f6b] raised exception: DatabaseError('server closed the connection unexpectedly\n\tThis probably means the server terminated abnormally\n\tbefore or while processing the request.\n',)
Traceback (most recent call last):
  File "/usr/local/lib/python2.6/dist-packages/celery/execute/trace.py", line 34, in trace
    return cls(states.SUCCESS, retval=fun(*args, **kwargs))
  File "/usr/local/lib/python2.6/dist-packages/celery/task/base.py", line 234, in __call__
    return self.run(*args, **kwargs)
  File "/usr/local/ast-tools/WebTools/../WebTools/controlpanel/tasks.py", line 54, in wrapper
    IP.objects.filter(ip=ip).update(busy="N")
  File "/usr/local/lib/python2.6/dist-packages/django/db/models/query.py", line 467, in update
    rows = query.get_compiler(self.db).execute_sql(None)
  File "/usr/local/lib/python2.6/dist-packages/django/db/models/sql/compiler.py", line 861, in execute_sql
    cursor = super(SQLUpdateCompiler, self).execute_sql(result_type)
  File "/usr/local/lib/python2.6/dist-packages/django/db/models/sql/compiler.py", line 727, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python2.6/dist-packages/django/db/backends/util.py", line 15, in execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python2.6/dist-packages/django/db/backends/postgresql_psycopg2/base.py", line 50, in execute
    return self.cursor.execute(query, args)
DatabaseError: server closed the connection unexpectedly
    This probably means the server terminated abnormally
    before or while processing the request.
```

Here's database log:

```
2011-04-14 18:26:37 CEST LOG:  connection received: host=127.0.0.1 port=40295
2011-04-14 18:26:37 CEST LOG:  connection authorized: user=ast database=astdb
2011-04-14 18:26:37 CEST LOG:  disconnection: session time: 0:00:00.044 user=ast database=astdb host=127.0.0.1 port=40295
2011-04-14 18:26:54 CEST LOG:  statement: UPDATE "djcelery_workerstate" SET "hostname" = E'ast-tool01', "last_heartbeat" = E'2011-04-14 18:26:52.081696' WHERE "djcelery_workerstate"."id" = 2 
2011-04-14 18:27:02 CEST LOG:  statement: INSERT INTO "djcelery_taskstate" ("state", "task_id", "name", "tstamp", "args", "kwargs", "eta", "expires", "result", "traceback", "runtime", "worker_id", "hidden") VALUES (E'FAILURE', E'352493f7-1fe5-4216-92aa-da44478233d2', E'WebTools.TestingReports.tasks.parseTestResultTaskv2', E'2011-04-14 18:27:00.068098', E'[u''/media/sdb/AST_TestLogs/testresults/'']', E'{}', NULL, NULL, E'DatabaseError(''server closed the connection unexpectedly\\n\\tThis probably means the server terminated abnormally\\n\\tbefore or while processing the request.\\n'',)', E'Traceback (most recent call last):
      File "/usr/local/lib/python2.6/dist-packages/celery/execute/trace.py", line 34, in trace
        return cls(states.SUCCESS, retval=fun(*args, **kwargs))
      File "/usr/local/lib/python2.6/dist-packages/celery/task/base.py", line 234, in __call__
        return self.run(*args, **kwargs)
      File "/usr/local/ast-tools/WebTools/../WebTools/TestingReports/tasks.py", line 45, in wrapper
        return func(*args, **kwargs)
      File "/usr/local/ast-tools/WebTools/../WebTools/TestingReports/tasks.py", line 243, in run
        tr.save()
      File "/usr/local/ast-tools/WebTools/../WebTools/TestingReports/models.py", line 60, in save
        super(TestReport, self).save(*args, **kwargs)
      File "/usr/local/lib/python2.6/dist-packages/django/db/models/base.py", line 434, in save
        self.save_base(using=using, force_insert=force_insert, force_update=force_update)
      File "/usr/local/lib/python2.6/dist-packages/django/db/models/base.py", line 527, in save_base
        result = manager._insert(values, return_id=update_pk, using=using)
      File "/usr/local/lib/python2.6/dist-packages/django/db/models/manager.py", line 195, in _insert
        return insert_query(self.model, values, **kwargs)
      File "/usr/local/lib/python2.6/dist-packages/django/db/models/query.py", line 1479, in insert_query
        return query.get_compiler(using=using).execute_sql(return_id)
      File "/usr/local/lib/python2.6/dist-packages/django/db/models/sql/compiler.py", line 783, in execute_sql
        cursor = super(SQLInsertCompiler, self).execute_sql(None)
      File "/usr/local/lib/python2.6/dist-packages/django/db/models/sql/compiler.py", line 727, in execute_sql
        cursor.execute(sql, params)
      File "/usr/local/lib/python2.6/dist-packages/django/db/backends/postgresql_psycopg2/base.py", line 50, in execute
        return self.cursor.execute(query, args)
    DatabaseError: server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
    ', NULL, 2, false)
```

Anybody knows what's going on ?
Thanks

**Comment by DylanYoung** [comment]
This is still an issue in 3.0.17. Just had Celery go down on 6 instances running Django 1.3.  Could be that a newer version of Django fixes it (none of our 1.6 instances went down).

**Comment by auvipy** [comment]
try to upgrade to 1.8

**Comment by DylanYoung** [comment]
If only that were as easy as you make it sound. Working on it though.

**Comment by auvipy** [comment]
it is not a good idea to run un-supported software frameworks.

**Comment by mrgaolei** [comment]
@ask This bug to reproduce on my server.

**Comment by Revocsid** [comment]
Same problem here with periodic tasks that need to fetch / update models stored in my PostgreSQL db. 

Django : 1.11.1
Psycopg2 : 2.7.1 
Celery : 4.0.2
PostgreSQL : 9.6.2

**Comment by goinnn** [comment]
Me too, is it related?

https://github.com/psycopg/psycopg2/issues/443

This will be fixed in the next psycopg2 release (2.7.2)

**Comment by dojeda** [comment]
I've stumbled on this exact problem. Difficult to pin down since I could not reproduce it at first.

To reproduce it, I created a dummy task that would query the database with an artificially long query. Something like
```
@shared_task
def dummy():
    result = list(SomeModel.objects.raw('SELECT id, pg_sleep(1) FROM some_table LIMIT 10'))
```

This task will connect with the database for 10 seconds.
On the other hand, I connected to my postgres database and severed the connection while the task is running with:
```
SELECT pid, pg_terminate_backend(pid), datid, client_addr, client_hostname, state FROM pg_stat_get_activity(NULL::integer) 
WHERE datid=(SELECT oid from pg_database where datname = 'django_db');
```
Where `django_db` is the database that I have for my Django application.

This procedure reproduced my problem and gave me something to work with. My best workaround is to add `django.db.utils.OperationalError` to the `autoretry_for` of my task. Since I had many tasks, and was already using a different base Task object, and did not want to change them all, I wrapped the `celery.shared_task` decorator:
```
import celery
def shared_task(*args, **kwargs):
    """Task decorator for robust tasks

    Robust tasks are tasks that will be retried 5 times if one of the
    exceptions in ...
    """
    kwargs['base'] = kwargs.get('base', None) or BaseTask  # BaseTask is my abstract base class that has some custom logging logic
    # Add the django.db.utils.OperationalError in the auto retry list of
    # exceptions. It seems that postgres is closing some connections and this
    # makes the task fail
    kwargs['autoretry_for'] = kwargs.get('autoretry_for', ()) + (django.db.utils.OperationalError,)
    if 'max_retries' not in kwargs:
        kwargs['max_retries'] = 5
    if 'countdown' not in kwargs and 'eta' not in kwargs:
        # TODO: the countdown might not be working on celery 4.1.0 ?
        # I haven't had any problems yet, but keep an eye on this issue:
        # https://github.com/celery/celery/issues/4221
        kwargs['countdown'] = 10
    return celery.shared_task(*args, **kwargs)
```
